### PR TITLE
WIP: Implement field indexes

### LINKED ISSUE #78: db_index parameter?
I miss the ability to mark fields as indexed. I don't know if this is a limitation of pypika. I need this feature to ensure that DB queries are decently fast. :)

### PR DESCRIPTION
**Pull request by florimondmanca**
## Description
- Add a new `index` boolean parameter to the base `Field` constructor.
- When `True`, an index is created for that field. This is implemented by adding a `CREATE INDEX` statement in the table creation query.
- If `safe=True`,  the index is only created if it doesn't exist already. For this we use a deterministic index name. Not supported on MySQL.

## Motivation and Context
Fixes #78.

If we want to implement multi-field indexes in the future, the index creation generation code may need to be refactored.

## How Has This Been Tested?
- I have changed `Tournament`'s `created` field to declare `index=True`
- New tests verify that a `CREATE INDEX` statement is added for the field when the schema is generated.

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

## TODO

- [x] Fix issues with `EXPLAIN` being database-dependant, or find a new way to verify an index was created.
- [x] Use the `safe` flag to only create indexes if they don't exist already.

### CODE REVIEW

**Comment by grigi** [comment]
Thanks for this Florimond. Generally the code looks good :+1:

The issue with the tests is due to either there not being a standard explain format #90 #91, or Tortoise not supporting basic driver capability reporting #86 

Ultimately I think BOTH should get solved, but either will unblock this :+1:

**Comment by florimondmanca** [comment]
Yes, the failing jobs are when using MySQL and PostgreSQL which do have a different syntax for the EXPLAIN statement than SQLite.

**Comment by grigi** [comment]
@florimondmanca Please review #92 to see if it would help unblock this?

**Comment by florimondmanca** [comment]
Yes it would: I'd mark the test with `@requireCapability(dialect="sqlite")` so it is skipped on MySQL and PostgreSQL. 😄

File: tortoise/backends/base/schema_generator.py [hunk 1]
```diff
@@ -83,13 +105,23 @@ def _get_table_sql(self, model, safe=True) -> dict:
...
         table_fields_string = ', '.join(fields_to_create)
         table_create_string = self.TABLE_CREATE_TEMPLATE.format(
             exists="IF NOT EXISTS " if safe else "",
             table_name=model._meta.table,
             fields=table_fields_string,
         )
 
+        for field_name in fields_with_index:
+            table_create_string += self.INDEX_CREATE_TEMPLATE.format(
+                index_name=self._generate_index_name(model, [field_name]),
```
Comment (florimondmanca) [inline tortoise/backends/base/schema_generator.py]: Note to self: need to add an `IF NOT EXISTS` check here if `safe == True`.

File: tortoise/backends/mysql/schema_generator.py [hunk 2]
```diff
@@ -17,9 +20,8 @@ def __init__(self, *args, **kwargs) -> None:
         )
 
         self.FIELD_TYPE_MAP.update({
-            fields.FloatField: 'DOUBLE',
-            fields.DatetimeField: 'DATETIME(6)',
-            fields.TextField: 'TEXT'
+            fields.FloatField: 'DOUBLE', fields.DatetimeField: 'DATETIME(6)', fields.TextField:
+            'TEXT'
```
Comment (florimondmanca) [inline tortoise/backends/mysql/schema_generator.py]: @grigi I turned on yapf in my text editor and this is how it reformatted the code. Looks strange to me, but I haven't used yapf yet so I don't know if this is normal?

Comment (grigi) [reply tortoise/backends/mysql/schema_generator.py]: We tried code formatters, both black and yapf, and was unsatisfied with either of them.
We decided it was better to stick with human formatted code, as it was much more readable.
Generally we are not too strict re style, flake8 checks for the worst offenders, and then it should be readable. 😐

**Comment by florimondmanca** [comment]
Actually, now that `.explain()` has been implemented, I was able to generalise the tests so they can run on all 3 databases.

However, I'm running into the fact that MySQL doesn't have an `IF NOT EXISTS` option for indexes. 😞  I've looked for ideas on how to work around that, but it's particularly nasty. The [most upvoted answers](https://dba.stackexchange.com/questions/24531/mysql-create-index-if-not-exists) use queries I'm not even sure we can run in the schema creation code.

Any clues @grigi ?

**Comment by florimondmanca** [comment]
Yes, my intuition is that there's something bigger that needs to be handled to solve this issue with MySQL. It seems using some sort of migration-based schema generation instead of "forcing" the schema to be created on every initialisation would allow us to just create the index once and not have to deal with this kind of issue.

Anyway, for now I will use the new capabilities API and note that MySQL doesn't support field indexes. Then if using MySQL, raise a warning during schema generation saying that we're skipping index creation.

**Comment by grigi** [comment]
Well, there is a step between just emitting code, and full on migrations. That is to synchronize. 
I think it is fine to emit the index SQL when `safe=False`, else warn that we don't support safely creating indexes in MySQL, so we ask the user to apply them manually for now.
You could emit the "dangerous" statement as part of the warning. or as a second warning.

**Comment by florimondmanca** [comment]
FYI, the build is currently failing due to an issue with `@requireCapability` which I've submitted a fix for in #98.

**Comment by florimondmanca** [comment]
… Tests pass! :tada: ✅

File: tortoise/backends/base/schema_generator.py [hunk 3]
```diff
@@ -83,13 +114,34 @@ def _get_table_sql(self, model, safe=True) -> dict:
...
             exists="IF NOT EXISTS " if safe else "",
             table_name=model._meta.table,
             fields=table_fields_string,
         )
 
+        # Indexes.
+        field_indexes_sqls = [
+            self._get_index_sql(model, [field_name], safe=safe) for field_name in fields_with_index
+        ]
+        if safe and not self.client.capabilities.safe_indexes:
```
Comment (grigi) [inline tortoise/backends/base/schema_generator.py]: :+1: Nice to have generic implementation based off capabilities

File: tortoise/contrib/test/__init__.py [hunk 4]
```diff
@@ -258,10 +261,28 @@ def requireCapability(connection_name: str = 'models', **conditions: Any):
...
-            for key, val in conditions.items():
-                if getattr(db.capabilities, key) != val:
-                    raise SkipTest('Capability {key} != {val}'.format(key=key, val=val))
+            checkCapability(connection_name, **conditions)
             return test_item(*args, **kwargs)
         return skip_wrapper
     return decorator
+
+
+def checkCapability(connection_name: str = 'models', **conditions: Any):
```
Comment (grigi) [inline tortoise/contrib/test/__init__.py]: The terminology is unexpected.
What this does is skip the test in the middle of it, if the capability is not set up.
And the only use-case is for testing capabilities when testing schema generation.
Makes me think this is not really necessary for being part of the API?

Comment (florimondmanca) [reply tortoise/contrib/test/__init__.py]: So, should I instead get the connection, check for capabilities and skip the test (essentially copying the behavior of `requireCapability` but not as a decorator) in these specific tests?

**Comment by florimondmanca** [comment]
> Can you please also look at making the `safe=True` optional parameters for the schema functions in `tortoise/utils.py` non-optional.

Done!

File: tortoise/contrib/test/__init__.py (hunk 4 above)
Comment (grigi) [reply tortoise/contrib/test/__init__.py]: I am in two minds about it. Primarily I'm unhappy with the naming, then I thought about if we really want to expose a piece of API that we probably don't need to expose.
So I'd say copy the code into the test module.

Comment (grigi) [reply tortoise/contrib/test/__init__.py]: I thought Travis was set up that way... Let me check.
It is like that. It only auto cancels a job that haven’t been started unfortunately.

**Comment by grigi** [comment]
Thanks for your good work. I released v0.11.3 :+1:

**Pull request merged**